c/tskit/trees.c

```diff
@@ -8145,3 +8165,339 @@ tsk_treeseq_extend_edges(const tsk_treeseq_t *self, int max_iter,
+        goto out;
+    }
+
+    nodes_time = tsk_malloc(num_nodes * sizeof(*nodes_time));
+    if (nodes_time == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    for (i = 0; i != num_nodes; i++) {
+        result[i] = TSK_NULL; /* nodes outside windows */
+        nodes_time[i].value = tables->nodes.time[i];
+        nodes_time[i].index = i;
+    }
+    qsort(nodes_time, (size_t) num_nodes, sizeof(*nodes_time), cmp_node_argsort);
+
+    for (i = 0, j = 0; i != *num_time_windows; i++) {
+        while (j != num_nodes && nodes_time[j].value < time_windows[i + 1]) {
+            if (nodes_time[j].value >= time_windows[i]) {
+                result[nodes_time[j].index] = (tsk_id_t) i; /* nodes inside windows */
+            }
+            j += 1;
+        }
+    }
+out:
+    tsk_safe_free(nodes_time);
+    return ret;
+}
+
+int
+tsk_treeseq_pair_coalescence_stat(const tsk_treeseq_t *self, tsk_size_t num_sample_sets,
+    const tsk_size_t *sample_set_sizes, const tsk_id_t *sample_sets,
+    tsk_size_t num_set_indexes, const tsk_id_t *set_indexes, tsk_size_t num_windows,
+    const double *windows, tsk_size_t num_time_windows, const double *time_windows,
+    tsk_flags_t options, double *result)
+{
+    int ret = 0;
+    double left, right, remaining_span, window_span;
+    tsk_id_t e, p, c, u, v, n, w, i, j, k, row, col, inp;
+    tsk_size_t total_samples;
+    int weight;
+    tsk_tree_position_t tree_pos;
+    const tsk_table_collection_t *tables = self->tables;
+    const tsk_size_t num_nodes = tables->nodes.num_rows;
+    const double sequence_length = tables->sequence_length;
+    tsk_id_t *nodes_sample_set = NULL;
+    tsk_id_t *nodes_parent = NULL;
+    tsk_id_t *nodes_time_window = NULL;
+    int *nodes_sample = NULL;
+    int *sample_count = NULL;
+    int *inside = NULL;
+    int *outside = NULL;
+    double *coalescing_pairs = NULL;
+    double *nodes_weight = NULL;
+
+    /* check inputs */
+    ret = tsk_treeseq_check_windows(self, num_windows, windows, options);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = check_time_windows(num_time_windows, time_windows);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = check_set_indexes(num_sample_sets, 2 * num_set_indexes, set_indexes);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_treeseq_check_sample_sets(
+        self, num_sample_sets, sample_set_sizes, sample_sets);
+    if (ret != 0) {
+        goto out;
+    }
+    /* TODO: set defaults? */
+
+    /* map nodes to sample sets */
+    nodes_sample_set = tsk_malloc(num_nodes * sizeof(*nodes_sample_set));
+    if (nodes_sample_set == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    ret = get_sample_set_index_map(self, num_sample_sets, sample_set_sizes, sample_sets,
+        &total_samples, nodes_sample_set);
+    if (ret != 0) {
+        goto out;
+    }
+
+    /* initialize internal state */
+    inp = (tsk_id_t) num_sample_sets;
```

Comment:
Why not use num_sample_sets in the malloc here rather than ``inp``? Saves a cast, and it's easier to read.

Reply:
Likewise for later memsets etc.